server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4338,6 +4359,28 @@ public void afterRefresh(boolean didRefresh) {
         }
     }
 
+    /**
+     * Switch the current engine to a read-only engine.
+     */
+    public void resetToReadOnlyEngine() throws IOException {
+        assert Thread.holdsLock(mutex) == false : "resetting engine under mutex";
+        assert getActiveOperationsCount() == OPERATIONS_BLOCKED
+            : "resetting engine without blocking operations; active operations are [" + getActiveOperationsCount() + ']';
+        var engineConfig = newEngineConfig(replicationTracker);
+        synchronized (engineMutex) {
+            verifyNotClosed();
+            IOUtils.close(currentEngineReference.get());
```

Comment:
shouldn't we do some error handling here?

Reply:
Good idea. If there are errors resetting the engine, I think we can fail the shard. That way it can recover again.